### PR TITLE
xwinglib updates for loading space objects

### PR DESCRIPTION
**Pull request by Dark-Visor**

### CODE REVIEW

File: code/mission/import/xwinglib.cpp [hunk 1]
```diff
@@ -74,6 +74,23 @@ struct xwi_flightgroup {
 };
 #pragma pack(pop)
 
+struct xwi_objectgroup {
```
Comment (Goober5000) [inline code/mission/import/xwinglib.cpp:76]: Move the `#pragma pack(pop)` line below the new struct.  The pack(push) and pack(pop) commands should enclose the structs.

File: code/mission/import/xwinglib.h [hunk 2]
```diff
@@ -138,6 +138,51 @@ enum class XWMObjective : short
...
+	oj_Orange_Crescent7,
+	oj_Orange_Crescent8,
+	oj_Death_Star
+};
+
+enum class XWMObjectFormation : short
+{
+	ojf_Flat,
+	ojf_Edge,
+	ojf_Broadside,
```
Comment (Goober5000) [inline code/mission/import/xwinglib.h]: Where did you get these values?  From the file format reference I see
```
00	Floor (X-Y plane)
01	Side (Y-Z plane)
02	Front (X-Z plane)
```

File: code/mission/import/xwinglib.h [hunk 3]
```diff
@@ -138,6 +138,51 @@ enum class XWMObjective : short
...
+	ojf_Scattered,
+	ojf_Destroy
```
Comment (Goober5000) [inline code/mission/import/xwinglib.h]: As I said in the DM:
> The mission goal will need to be split off and stored in a separate field in the friendly struct

So `ojf_Destroy` should not be included here

File: code/mission/import/xwinglib.h [hunk 4]
```diff
@@ -210,6 +255,22 @@ class XWMObject
 {
 public:
 
+	std::string designation;
+	std::string cargo;
+	std::string specialCargo;
+
+	int specialObjectNumber;
```
Comment (Goober5000) [inline code/mission/import/xwinglib.h]: Since these fields aren't actually used, then you might as well take them out.

File: code/mission/import/xwinglib.h [hunk 5]
```diff
@@ -210,6 +255,22 @@ class XWMObject
...
+
+	XWMObjectType objectType;
+
+	XWMCraftIFF objectIFF; // shares with FlightGroup
```
Comment (Goober5000) [inline code/mission/import/xwinglib.h]: Since mines are always hostile, and IFF is not relevant to other objects, you might as well take the IFF out too.

File: code/mission/import/xwinglib.h [hunk 6]
```diff
@@ -210,6 +255,22 @@ class XWMObject
...
+
+	XWMObjectFormation formation;
```
Comment (Goober5000) [inline code/mission/import/xwinglib.h:276]: You'll need to add a separate enum (survive/destroyed/neither) to handle the goal.

File: code/mission/import/xwinglib.cpp [hunk 7]
```diff
@@ -582,5 +599,161 @@ bool XWingMission::load(XWingMission *m, const char *data)
 		m->flightgroups.push_back(*nfg);
 	}
 
+	for (int n = 0; n < h->number_of_objects; n++) {
+		xwi_objectgroup *oj =
+			(xwi_objectgroup*)(data + sizeof(xwi_header) + sizeof(xwi_flightgroup) + sizeof(xwi_objectgroup) * n);
```
Comment (Goober5000) [inline code/mission/import/xwinglib.cpp]: The `sizeof(xwi_flightgroup)` should be multiplied by `h->number_of_flight_groups` because the object data starts after *all* the flight groups have been parsed.

File: code/mission/import/xwinglib.cpp [hunk 8]
```diff
@@ -582,5 +599,161 @@ bool XWingMission::load(XWingMission *m, const char *data)
...
+		XWMObject noj_buf;
+		XWMObject *noj = &noj_buf;
+
+		noj->designation = oj->designation;
+		noj->cargo = oj->cargo;
+		noj->specialCargo = oj->special_cargo;
+		noj->specialObjectNumber = oj->special_object_number;
+
+		switch (oj->object_type) {
```
Comment (Goober5000) [inline code/mission/import/xwinglib.cpp:609]: Look at the `CraftType` list in the file format.  Mine1 is hexadecimal 12 (decimal 18) and the other object types likewise need their indexes fixed.

File: code/mission/import/xwinglib.cpp [hunk 9]
```diff
@@ -582,5 +599,161 @@ bool XWingMission::load(XWingMission *m, const char *data)
...
+				break;
+			case 2:
+				noj->objectIFF = XWMCraftIFF::iff_imperial;
+				break;
+			case 3:
+				noj->objectIFF = XWMCraftIFF::iff_neutral;
+				break;
+		}
+
+		switch (oj->object_formation) {
```
Comment (Goober5000) [inline code/mission/import/xwinglib.cpp]: See the other review comments about object formation.  Note the special handling of bits 3 and 4: these are handled *independently*, so they need special bitwise checks.

File: code/mission/import/xwinglib.cpp [hunk 10]
```diff
@@ -582,5 +599,161 @@ bool XWingMission::load(XWingMission *m, const char *data)
...
+		}
+
+		noj->numberOfObjects = oj->number_of_objects;
+
+		noj->object_x = oj->object_x / 160.0f;
+		noj->object_y = oj->object_y / 160.0f;
+		noj->object_z = oj->object_z / 160.0f;
+		noj->object_yaw = oj->object_yaw / 160.0f;
+		noj->object_pitch = oj->object_pitch / 160.0f;
+		noj->object_roll = oj->object_roll / 160.0f;
```
Comment (Goober5000) [inline code/mission/import/xwinglib.cpp]: The yaw/pitch/roll should not be divided by 160.  The pitch, though, should subtract 90.  Again, consult the file format reference.

File: code/mission/import/xwingmissionparse.cpp [hunk 11]
```diff
@@ -623,6 +623,144 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
 	}
 }
 
+void parse_xwi_objectgroup(mission *pm, const XWingMission *xwim, const XWMObject *oj)
```
Comment (Goober5000) [inline code/mission/import/xwingmissionparse.cpp]: Split off the xwingmissionparse changes into a separate branch; they should not be part of this PR.  We'll cross that bridge when we come to it.

**Review by Goober5000 — CHANGES_REQUESTED** [review]

File: code/mission/import/xwinglib.h (hunk 2 above)
Comment (Dark-Visor) [reply code/mission/import/xwinglib.h]: ah I was taking those from the older (archived) reference. I'll update

File: code/mission/import/xwinglib.h [hunk 12]
```diff
@@ -138,6 +138,67 @@ enum class XWMObjective : short
...
+	ojf_SideYZ,
+	ojf_FrontXZ,
+	ojf_Scattered // may be buggy - undefined locations
+};
+
+enum class XWMObjectGoal : short
+{
+	ojg_Survive,
+	ojg_Destroyed,
+	ojg_Neither
```
Comment (Goober5000) [inline code/mission/import/xwinglib.h]: Put `ojg_Neither` as the first item so it is assigned the value 0

File: code/mission/import/xwinglib.cpp [hunk 13]
```diff
@@ -72,6 +72,20 @@ struct xwi_flightgroup {
 	short primary_target;
 	short secondary_target;
 };
+
+struct xwi_objectgroup {
```
Comment (Goober5000) [inline code/mission/import/xwinglib.cpp:76]: The `xwi_objectgroup` struct must exactly match the file format reference, even if the `XWMObject` struct has more fields or fewer fields.  That's why I called `XWMObject` the "friendly" struct, because it is friendlier to what the user might do with it.  By contrast, you can't add or remove fields to/from `xwi_objectgroup` because the bytes will be misaligned.

File: code/mission/import/xwinglib.cpp [hunk 14]
```diff
@@ -582,5 +596,186 @@ bool XWingMission::load(XWingMission *m, const char *data)
...
+				noj->objectType = XWMObjectType::oj_Training_Platform11;
+				break;
+			case 69:
+				noj->objectType = XWMObjectType::oj_Training_Platform12;
+				break;
+			default:
+				return false;
+		}
+
+		switch (oj->object_formation) {
```
Comment (Goober5000) [inline code/mission/import/xwinglib.cpp]: The formation code here will need to mask off the non-relevant bytes.

File: code/mission/import/xwinglib.cpp [hunk 15]
```diff
@@ -582,5 +596,186 @@ bool XWingMission::load(XWingMission *m, const char *data)
...
+				noj->formation = XWMObjectFormation::ojf_FrontXZ;
+				break;
+			case 3:
+				noj->formation = XWMObjectFormation::ojf_Scattered;
+				break;
+			default:
+				return false;
+		}
+
+		switch (oj->object_goal) {
```
Comment (Goober5000) [inline code/mission/import/xwinglib.cpp]: There is no `object_goal` field in the byte structure (see the comment about `xwi_objectgroup`).  Instead `noj->objectGoal` must be set using the bit checks.

**Review by Goober5000 — CHANGES_REQUESTED** [review]

**Review by Goober5000 — CHANGES_REQUESTED** [review]
Excellent work.  This is nearly ready.

There's one more thing we should do.  If the object is any type of training platform, we should skip over the regular formation handling and just assign a formation of FloorXY and a goal of Neither.  According to the file format reference, the `object_formation` field controls which guns are present on the training platform, and how many seconds are on the clock for training missions.  This would be annoying and tedious to code, for not much gain.  Put a TODO there, and we can come back to it in the future if anyone desperately wants to import training missions.

**Review by Goober5000 — APPROVED** [review]
Fantastic work 👍
